### PR TITLE
Optimize handleCheckState()

### PR DESCRIPTION
**Pull request by Methuselah96**

### CODE REVIEW

**Comment by warmhug** [comment]
Too much change! Attention to potential problems, need to be verified @yesmeck

**Comment by yesmeck** [comment]
Hello @Methuselah96 could you describe why these changes optimized handleCheckState? And could you write a test plan for these changes?

**Comment by Methuselah96** [comment]
Yes. Before, handleCheckState() found the children and parents of the tree node through string manipulation (String#split) which is extremely inefficient. These changes now store the children positions and the parent position of each node in the tree node's state. That way, handleCheckSate() can now just loop through the tree node's children nodes recursively and update their checked state and then loop through the tree node's parent nodes recursively and update their checked state. Using the big-data example with X: 1000, Y:2, Z:1 I was able to cut certain check events from taking about 10 seconds to about 2 seconds (all of those 2 secs are now because it has to rerender the entire tree). What do you mean by a "test plan"? Do you mean you want me to write another test in Tree.spec.js? I would assume that's already being tested.

**Comment by yesmeck** [comment]
A "test plan" means how can I verify your PR, you've mentioned using the big-data example `with X: 1000, Y:2, Z:1`, and you said "I was able to cut certain check events from taking about 10 seconds to about 2 seconds", how do you measure the time?

**Comment by Methuselah96** [comment]
Yeah, so I noticed that the current algorithm for `handleCheckState` is especially slow for checking boxes when there are a lot of other check boxes checked. So, with `X: 1000, Y:2, Z:1`, I check the first 10 top level folders and files (the `0-0` and `0-1` folders and the `0-2` through `0-9` files). Then I expand `0-0` and `0-0-0` and uncheck `0-0-0-0`. I open up the Chrome DevTools performance profiler, start recording performance, and then check `0-0-0-0`. Then I stop the performance profiler and look at how long the `onClick` method took. On my computer, I notice a drop from about 10 seconds to about 2 seconds.

**Comment by afc163** [comment]
Seem that the code is much more clean.

File: tests/util.test.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { arraysEqual } from '../src/util';
+
+test('[0, 1] equals [0, 1]', () => {
+  const array1 = [0, 1];
+  const array2 = [0, 1];
+  expect(arraysEqual(array1, array2)).toBe(true);
+});
+
+test('[0, 1] does not equal [1, 0]', () => {
```
Comment (benjycui) [inline tests/util.test.js]: This test case is wrong.

**Comment by Methuselah96** [comment]
@benjycui @yesmeck Fixed test case.

**Pull request merged**